File: server/utils/leagues/permissions.ts
export type LeagueRole = 'OWNER' | 'MODERATOR' | 'MEMBER'

// Rename, regenerate the join code, kick members.
export function canManageLeague(role: LeagueRole | null | undefined): boolean {
  return role === 'OWNER' || role === 'MODERATOR'
}

export function canKick(actor: LeagueRole | null | undefined, target: LeagueRole | null | undefined): boolean {
  if (actor === 'OWNER') return target === 'MODERATOR' || target === 'MEMBER'
  if (actor === 'MODERATOR') return target === 'MEMBER'
  return false
}

// Members ask a mod for the code; that keeps "regenerate code" a real
// invite-revocation tool instead of a no-op.
export function canSeeJoinCode(role: LeagueRole | null | undefined): boolean {
  return role === 'OWNER' || role === 'MODERATOR'
}

File: apps/mobile-flutter/app/lib/leagues/permissions.dart
/// Client mirror of `apps/web-nuxt/server/utils/leagues/permissions.ts` plus the
/// owner-only role rules the routes enforce. The server is authoritative; this
/// exists so the UI never offers an action that is going to 403.
library;

import '../api/models.gen.dart';

bool canManageLeague(RoleValue? role) =>
    role == RoleValue.owner || role == RoleValue.moderator;

bool canKick(RoleValue? actor, RoleValue? target) {
  if (actor == RoleValue.owner) {
    return target == RoleValue.moderator || target == RoleValue.member;
  }
  if (actor == RoleValue.moderator) return target == RoleValue.member;
  return false;
}

bool canSeeJoinCode(RoleValue? role) => canManageLeague(role);

/// `PUT /api/leagues/[id]/members/[userId]` requires OWNER.
bool canChangeRole(RoleValue? actor, RoleValue? target) =>
    actor == RoleValue.owner &&
    (target == RoleValue.member || target == RoleValue.moderator);

/// Ownership moves through the transfer endpoint, owner only.
bool canTransferOwnership(RoleValue? actor, RoleValue? target) =>
    actor == RoleValue.owner && target != RoleValue.owner;
